src/core/internal/hash.d

```diff
@@ -44,38 +172,51 @@ size_t hashOf(T)(auto ref T val, size_t seed = 0) if (is(T == enum))
     }
 }
 
-//CTFE ready (depends on base type). Can be merged with dynamic array hash
-size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && __traits(isStaticArray, T))
+//CTFE ready (depends on base type).
+size_t hashOf(T)(scope const auto ref T val, size_t seed = 0)
+if (!is(T == enum) && __traits(isStaticArray, T) && canBitwiseHash!T)
 {
-    size_t cur_hash = seed;
-    foreach (ref cur; val)
+    static if (T.length == 0)
+        return seed;
+    else static if (T.length == 1)
+        return hashOf(val[0], seed);
+    else static if (is(typeof(toUbyte(val)) == const(ubyte)[]))
+        return bytesHashAlignedBy!T(toUbyte(val), seed);
+    else
     {
-        cur_hash = hashOf(cur, cur_hash);
+        assert(!__ctfe, "unable to compute hash of "~T.stringof~" at compile time");
+        const(ubyte)[] bytes = (() @trusted => (cast(const(ubyte)*)&val)[0 .. T.sizeof])();
+        return bytesHashAlignedBy!T(bytes, seed);
     }
-    return cur_hash;
 }
 
-//dynamic array hash
-size_t hashOf(T)(T val, size_t seed = 0)
-if (!is(T == enum) && !is(T : typeof(null)) && is(T S: S[]) && !__traits(isStaticArray, T)
-    && !is(T == struct) && !is(T == class) && !is(T == union))
+//CTFE ready (depends on base type).
+size_t hashOf(T)(auto ref T val, size_t seed = 0)
+if (!is(T == enum) && __traits(isStaticArray, T) && !canBitwiseHash!T)
 {
-    alias ElementType = typeof(val[0]);
-    static if (is(ElementType == interface) || is(ElementType == class) ||
-                   (is(ElementType == struct) && !isNonReference!ElementType) ||
-                   ((is(ElementType == struct) || is(ElementType == union))
-                       && is(typeof(val[0].toHash()) == size_t)))
-    //class or interface array or struct array with toHash(); CTFE depend on toHash() method
-    //also do this for arrays of structs whose hashes would be calculated in a memberwise fashion
-    {
-        size_t hash = seed;
-        foreach (o; val)
+    static if (T.length == 0)
+        return seed;
+    else static if (T.length == 1)
+        return hashOf(val[0], seed);
+    else
+    {
+        size_t result = seed;
+        foreach (ref cur; val)
         {
-            hash = hashOf(o, hash);
+            result = hashOf(cur, result);
         }
-        return hash;
+        return result;
```

Comment:
If you group them `canBitwiseHash` only needs to be evaluated once (and it might also be a nice logical grouping).

Reply:
Doesn't the compiler cache the result of template instantiation?

Reply:
Knowing that I'm using `bytesHash` is what lets me know that it can be `scope const`. Is being able to infer that worth the tradeoff of having an extra overload?

Reply:
> Doesn't the compiler cache the result of template instantiation?

Oh it does: https://run.dlang.io/is/y9PxzM

> Knowing that I'm using bytesHash is what lets me know that it can be scope const. Is being able to infer that worth the tradeoff of having an extra overload?

It probably is.